### PR TITLE
Guard LLM-generated sql_query rules

### LINKED ISSUE #1123: [FEATURE]: Guard results of AI-assisted rules generation
### Is there an existing issue for this?

- [x] I have searched the existing issues

### Problem statement

Output is not capped, and the generated `sql_query` rule is not validated against malicious SQL. 

### Proposed Solution

Add `max_tokens` to cap input length.
is_sql_query_safe() on any sql_query in the LLM output before returning

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by aarushisingh04**
## Changes

adds SQL safety validation to both LLM-assisted rule generation paths so that any `sql_query` rule containing unsafe SQL (DML/DDL) is dropped before being returned to the caller.

- added `_filter_unsafe_sql_rules()` in `llm/llm_core.py` : parses the generated JSON rules array and drops any rule whose `check.function` is `sql_query` and whose `query` argument fails `is_sql_query_safe()`, logging a warning with the sanitized query string (CWE-117)
- hooked it into `DspyRuleGeneration.forward()` and `DspyRuleUsingDataStats.forward()` via an `else` clause on the existing JSON-validation block; `DspyRuleGenerationWithSchemaInference` inherits the fix automatically since it delegates to `DspyRuleGeneration`
> note: `max_tokens` was already present in `LLMModelConfig` and forwarded to `dspy.LM()`, it caps completion length, not prompt input, which is sufficient for the issue's intent of bounding unbounded LLM output.
- fixed training example for ai-assisted rules generation

### Linked issues

resolves #1123

### Tests

- [ ] manually tested
- [x] added unit tests
- [ ] added integration tests
- [ ] added end-to-end tests
- [ ] added performance tests

19 unit tests in `tests/unit/test_llm_core.py` covering: safe SELECT queries pass through, unsafe DROP/DELETE/INSERT/TRUNCATE rules are dropped, non-`sql_query` rules are unaffected, mixed safe+unsafe batches, null/missing `query` argument, non-array JSON input, malformed rule entries, invalid JSON passthrough, and warning log emission.

### Documentation and Demos

- [ ] added/updated demos
- [ ] added/updated docs
- [ ] added/updated agent skills

### CODE REVIEW

File: src/databricks/labs/dqx/llm/llm_core.py [hunk 1]
```diff
@@ -7,12 +7,62 @@
...
+        check = rule.get("check", {})
+        if not isinstance(check, dict):
+            continue
+
+        if check.get("function") == "sql_query":
+            arguments = check.get("arguments", {})
+            if not isinstance(arguments, dict):
+                continue
+
+            query = arguments.get("query") or ""
```
Comment (mwojtyczka) [inline src/databricks/labs/dqx/llm/llm_core.py]: **A non-string `query` will crash rule generation instead of being filtered.** `arguments.get("query") or ""` keeps a non-string truthy value as-is, and `is_sql_query_safe()` immediately does `re.sub(r"\s+", " ", query)` → `TypeError: expected string or bytes-like object` for an `int`/`list`/`dict`. This function doesn't catch it, and the caller's `try/except` only wraps `json.loads` (the `_filter_unsafe_sql_rules` call is in the `else` branch, outside the try) — so malformed LLM output like `{"query": 123}` raises an unhandled `TypeError` straight out of `forward()` and aborts generation. That's exactly the bad-LLM-output case this guard exists for, and it's the one input shape not already handled (non-dict rule/check/arguments and null/missing query all are). Suggest: `q = arguments.get("query"); if not isinstance(q, str): q = ""` (or treat non-str as unsafe and drop), plus a `query: 123` regression test. (Reproduced locally.)

File: src/databricks/labs/dqx/llm/llm_core.py [hunk 2]
```diff
@@ -147,13 +197,15 @@ def forward(
...
-        # Validate JSON output
+        # Validate JSON output and filter unsafe sql_query rules
         if result.quality_rules:
             try:
                 json.loads(result.quality_rules)
             except json.JSONDecodeError as e:
                 logger.warning(f"Generated invalid JSON: {e}. Returning empty rules.")
                 result.quality_rules = "[]"
+            else:
+                result.quality_rules = _filter_unsafe_sql_rules(result.quality_rules)
```
Comment (mwojtyczka) [inline src/databricks/labs/dqx/llm/llm_core.py]: Minor: the `try/except/else` control flow is correct (the `else` runs only when `json.loads` succeeds, which is the intent). But `result.quality_rules` is now parsed twice — once here just to validate, then again inside `_filter_unsafe_sql_rules`. You could parse once and pass the parsed `list[dict]` into the filter (having it return the filtered list, re-serialized by the caller), avoiding the redundant `json.loads`. Not a correctness issue — same remark applies to the identical block in `DspyRuleUsingDataStats.forward`

File: tests/unit/test_llm_core.py [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+
+
+def test_filter_unsafe_sql_rules_null_query_kept():
+    # null query defaults to "" (safe); engine will reject it later via MissingParameterError
+    rules = [{"check": {"function": "sql_query", "arguments": {"query": None}}}]
+    result = json.loads(_filter_unsafe_sql_rules(json.dumps(rules)))
+    assert len(result) == 1
+
+
+def test_filter_unsafe_sql_rules_missing_query_key_kept():
```
Comment (mwojtyczka) [inline tests/unit/test_llm_core.py]: Good coverage overall (happy paths, drop/keep, malformed rule/check, null & missing `query`, invalid-JSON passthrough, plus integration for two of the generators). A few gaps worth closing, and notably they overlap the edge cases most likely to break:

1. **Non-string `query` (e.g. `{"query": 123}`)** — not tested, and it's exactly the input that currently raises an unhandled `TypeError` (see my comment on `llm_core.py:55`). Add this test together with the type guard so it's a regression test, asserting the rule is dropped (or kept as `""`) rather than crashing.
2. **`arguments` not a dict** (e.g. `"arguments": "DROP TABLE x"`) — the `if not isinstance(arguments, dict): continue` branch has no test.
3. **`DspyRuleGenerationWithSchemaInference.forward`** — the third public generation path has no test. It inherits the guard via delegation, but a test would catch a future regression in that path.
4. **Outer `if result.quality_rules` falsy case** (empty string / `None`) isn't exercised at the integration level — the "not json" test covers a non-empty invalid value, not an empty/None one.

**Review by mwojtyczka — CHANGES_REQUESTED** [review]
Thank you for the PR. Generally looking good. Left some comments to address.

**Comment by aarushisingh04** [comment]
@mwojtyczka 

appreciate the thorough review! addressed all three points in the latest push:

- added `isinstance(query, str)` guard, non-string query is now dropped with a warning instead of crashing
- refactored `_filter_unsafe_sql_rules` to accept a pre-parsed list, eliminating the double-parse
- added missing tests: non-string query, non-dict arguments, `DspyRuleGenerationWithSchemaInference` path, and empty/None `quality_rules` at the integration level

**Review by mwojtyczka — APPROVED** [review]
LGTM
